src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/HMPPSAuthService.kt

```diff
@@ -104,15 +117,52 @@ class HMPPSAuthService(
         hmppsAuthApiClient.get(detailUrl)
           .retrieve()
           .bodyToMono(UserDetailResponse::class.java)
+          .withRetryPolicy()
           .map { it.name.substringBefore(' ') },
         hmppsAuthApiClient.get(emailUrl)
           .retrieve()
           .onStatus({ it.equals(HttpStatus.NO_CONTENT) }, { Mono.error(UnverifiedEmailException()) })
           .bodyToMono(UserEmailResponse::class.java)
+          .withRetryPolicy()
           .map { it.email }
       )
         .map { UserDetail(it.t1, it.t2) }
         .block()
     }
   }
+
+  fun <T> Flux<T>.withRetryPolicy(): Flux<T> {
+    return this
+      .retryWhen(
+        Retry.max(maxRetryAttempts)
+          .filter { isTimeoutException(it) }
+          .doBeforeRetry { logRetrySignal(it) }
+      )
+  }
+
+  fun <T> Mono<T>.withRetryPolicy(): Mono<T> {
+    return this
+      .retryWhen(
+        Retry.max(maxRetryAttempts)
+          .filter { isTimeoutException(it) }
+          .doBeforeRetry { logRetrySignal(it) }
+      )
+  }
+
+  fun isTimeoutException(it: Throwable): Boolean {
+    // Timeout for NO_RESPONSE is wrapped in a WebClientRequestException
+    return it is ReadTimeoutException || it is ConnectTimeoutException ||
+      it.cause is ReadTimeoutException || it.cause is ConnectTimeoutException
```

Comment:
Do we need to go into causes?

Reply:
Asking because if we do, it's a recursive data structure, so ... is it sufficient to stop at level 1?

Reply:
In the case of NO_RESPONSE (as apposed to delayed) spring does seem to wrap it with a WebClientRequestException so we do need to look inside.